src/backups.py

```diff
@@ -0,0 +1,307 @@
+
+        # Check if this unit can perform backup
+        can_unit_perform_backup, validation_message = self._can_unit_perform_backup()
+        if not can_unit_perform_backup:
+            logger.warning(validation_message)
+            event.fail(validation_message)
+            return
+
+        # Test uploading metadata to S3 to test credentials before backup
+        juju_version = JujuVersion.from_environ()
+        metadata = f"""Date Backup Requested: {datetime_backup_requested}
+Model Name: {self.model.name}
+Application Name: {self.model.app.name}
+Unit Name: {self.charm.unit.name}
+Juju Version: {str(juju_version)}
+"""
+
+        if not upload_content_to_s3(
+            metadata,
+            s3_parameters["bucket"],
+            f"{s3_directory}.metadata",
+            s3_parameters["region"],
+            s3_parameters["endpoint"],
+            s3_parameters["access-key"],
+            s3_parameters["secret-key"],
+        ):
+            event.fail("Failed to upload metadata to provided S3")
+            return
+
+        # Run operations to prepare for the backup
+        success, error_message = self._pre_backup()
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+            return
+
+        # Perform the backup
+        success, error_message = self._backup(
+            s3_parameters["bucket"],
+            s3_directory,
+            s3_parameters["endpoint"],
+            s3_parameters["region"],
+            s3_parameters["access-key"],
+            s3_parameters["secret-key"],
+        )
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+
+            success, error_message = self._post_backup()
+            if not success:
+                logger.warning(error_message)
+
+            return
+
+        # Run operations to clean up after the backup
+        success, error_message = self._post_backup()
+        if not success:
+            logger.warning(error_message)
+            event.fail(error_message)
+            return
+
+        event.set_results(
+            {
+                "backup-id": datetime_backup_requested,
+            }
+        )
+
+    def _retrieve_s3_parameters(self) -> Tuple[Dict, List[str]]:
+        """Retrieve S3 parameters from the S3 integrator relation."""
+        s3_parameters = self.s3_integrator.get_s3_connection_info()
+
+        required_parameters = [
+            "bucket",
+            "access-key",
+            "secret-key",
+        ]
+        missing_required_parameters = [
+            param for param in required_parameters if param not in s3_parameters
+        ]
+        if missing_required_parameters:
+            logger.warning(
+                f"Missing required S3 parameters in relation with S3 integrator: {missing_required_parameters}"
+            )
+            return {}, missing_required_parameters
```

Comment:
Normally, I would agree, but the reason why we have 2 return values here is because I would like to use the `missing_required_parameters` to compile the action failure message (would like to avoid using the exception to retrieve this)